keps/sig-network/20200204-cni-verification-rearchitecture.md

```diff
@@ -0,0 +1,579 @@
+that the NetworkPolicy tests *can* pass on *some* provider.  As part of this proposal, we propose commiting an annotation to the existing `network_policy.go` code which states, in clear and simple terms, what environment the `network_policy.go` test suite was run in, the last time which it was committed and passed.  It's also acceptable to commit this as a Markdown file in the documentation.
+ 
+There may be other, better ways of doing this.  Running an upstream validation job of these tests as a weekly PROW job, for example, would be a good way to make sure that these tests don't regress in the future.  This comes at the cost of coupling a job to an external CNI provider, so its not being explicitly suggested.
+
+## Other Improvement Ideas
+
+These may be included in this proposal, but as of now, aren't detailed yet.
+
+### Node specific policy validation (Contributed by Sedef Saavas)
+
+A wider range of scenarios that may be tested can be seen in the figure below.
+![](test-scenarios.png)
+ In this figure, inter/intra-namespace tests on the same node and inter-intra namespace tests on different nodes are demonstrated.
+Another important network-policy-test case is testing host-network only containers. C4, C5, C8, and C9 are host-network only containers.
+
+### Ensuring that large policy stacks evaluate correctly
+
+Right now the coverage of Policy stacks is rudimentary, we may want to test for a large number(i.e. 10) of policies, stacked, depending on whether we think this may be a bug source for providers.
+
+## Alternative solutions to this proposal
+ 
+#### Keeping the tests as they are and fixing them one by one
+ 
+We could simply audit existing tests for completeness, and one-by-one, add new test coverage where it is lacking.  This may be feasible for the 23 tests we currently have, but it would be likely to bit-rot over time, and not solve the extensibility or debuggability problems.
+ 
+#### Building a framework for NetworkPolicy evaluation
+ 
+In this proposal, we've avoided suggesting a complex framework that could generate large numbers of services and pods, and large permutations of scenarios.
+However, it should be noted that such a framework might be useful in testing performance at larger scales, and comparing CNI providers with one another. Such a framework could easily be adopted to cover the minimal needs of the NetworkPplicy implemetnation in core Kubernetes, so it might be an interesting initiative to work on.  Such an initiative might fall on the shoulders of another Sig, related to performance or scale.  Since NetworkPolicies have many easy-to-address
+problems which are important as they stand, we avoid going down this rat-hole, for now.
+ 
+That said, the work proposed here might be a first step towared a more generic CNI testing model.
+
+#### Have the CNI organization create such tests
```

Comment:
Also, NetworkPolicy is not part of the CNI spec (though it is traditionally implemented by CNI plugins).

Reply:
yep agree